x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java

```diff
@@ -1572,15 +1579,77 @@ static void writeAlias(PlanStreamOutput out, Alias alias) throws IOException {
     // -- Expressions (other)
 
     static Literal readLiteral(PlanStreamInput in) throws IOException {
-        return new Literal(in.readSource(), in.readGenericValue(), in.dataTypeFromTypeName(in.readString()));
+        return new Literal(
+            in.readSource(),
+            in.readGenericValue(),
+            in.dataTypeFromTypeName(in.readString()),
+            PlanNamedTypes::mapToLiteralValue
+        );
     }
 
     static void writeLiteral(PlanStreamOutput out, Literal literal) throws IOException {
         out.writeNoSource();
-        out.writeGenericValue(literal.value());
+        out.writeGenericValue(mapFromLiteralValue(out, literal.dataType(), literal.value()));
         out.writeString(literal.dataType().typeName());
     }
 
+    /**
+     * Not all literal values are currently supported in StreamInput/StreamOutput as generic values.
+     * This mapper allows for addition of new and interesting values without (yet) adding to StreamInput/Output.
+     * This makes the most sense during the pre-GA version of ESQL. When we get near GA we might want to push this down.
+     * <p>
+     * For the spatial point type support we need to care about the fact that 8.12.0 uses encoded longs for serializing
+     * while 8.13 uses WKB. We detect if the cluster has 8.12 using an existing known TransportVersion that was added for 8.13.
+     * We do not need more fine-grained checks at this point since ESQL is not supported in Serverless.
+     */
+    private static Object mapFromLiteralValue(PlanStreamOutput out, DataType dataType, Object value) {
+        if (dataType == GEO_POINT || dataType == CARTESIAN_POINT) {
+            // Picked an existing transport version that is clearly in between the doc-values and WKB versions of the point literals
+            // This simplification works since ESQL is not supported in serverless, so we're only differentiating between 8.12 and 8.13
+            if (out.getTransportVersion().before(TransportVersions.ESQL_PROFILE)) {
+                if (value instanceof List<?> list) {
+                    return list.stream().map(v -> mapFromLiteralValue(out, dataType, v)).toList();
+                }
+                if (value instanceof BytesRef wkb) {
```

Comment:
We should throw an exception if the value is not a BytesRef? I would either cast it directly so we throw a CastClassException or add an else statement? I think I prefer the first one as the error will be clear.

Reply:
Done.